### ISSUE TITLE
[0.50.0-SNAPSHOT] Artillery causes rapid slowdown followed by Out of Memory error.

### ISSUE THREAD

**Issue by Tzahr** [MegaMek/megamek#5763]
### Environment
14:37:01,282 INFO  [megamek.MegaMek] {main}
megamek.MegaMek.initializeLogging(MegaMek.java:143) - Starting MegaMek v0.50.0-SNAPSHOT
    Build Date: 2024-07-21T01:22:35.706888084
    Today: 2024-07-21
    Origin Project: MegaMek
    Java Vendor: Eclipse Adoptium
    Java Version: 21.0.3
    Platform: Windows 11 10.0 (amd64)
    System Locale: en_GB
    Total memory available to MegaMek: 2 GB
    MM Code Revision: 8bd3a4528070b207d5803ad74a5a4a6758b54ff7
    MML Code Revision: Unknown
    MHQ Code Revision: Unknown

### Description
When employing artillery, after 2-4 rounds, MegaMek slows rapidly around the resolution of the artillery phase, and then quickly becomes unresponsive afterwards, and gives an Out of Memory error. 

Tested this with both Princess- and player-controlled on- and off-board artillery, same results.

### Files
![image](https://github.com/user-attachments/assets/dd4251be-70b4-4498-ac1e-dfef2ffd1a0b)
[megamek.log](https://github.com/user-attachments/files/16324501/megamek.log)

**Comment by repligator** [comment]
Build Date: 2024-07-21T11:23:23.344653229
Today: 2024-07-21
Origin Project: MekHQ
Java Vendor: Ubuntu
Java Version: 17.0.11
Platform: Linux 6.5.0-44-generic (amd64)
System Locale: en_US
Total memory available to MegaMek: 4 GB
MM Code Revision: Unknown
MML Code Revision: Unknown
MHQ Code Revision: Unknown

Same/similar issue. Game starts chugging like it's running a Division vs Division instead of Lance vs Lance. I don't get an actual OOM error though. I'm running the same company from MHQ that I was yesterday. I had no issues yesterday, but I do today.

[megamek (Lockup1).log](https://github.com/user-attachments/files/16325102/megamek.Lockup1.log)
[megamek (Lockup2).log](https://github.com/user-attachments/files/16325103/megamek.Lockup2.log)
[megamek (Lockup3).log](https://github.com/user-attachments/files/16325104/megamek.Lockup3.log)
[gamelog.zip](https://github.com/user-attachments/files/16325107/gamelog.zip)
[autosave.sav.gz](https://github.com/user-attachments/files/16325109/autosave.sav.gz)
[Round-2-autosave_2024-07-21_11-10-13.sav.gz](https://github.com/user-attachments/files/16325110/Round-2-autosave_2024-07-21_11-10-13.sav.gz)
[Round-3-autosave_2024-07-21_11-13-22.sav.gz](https://github.com/user-attachments/files/16325112/Round-3-autosave_2024-07-21_11-13-22.sav.gz)
[Round-4-autosave_2024-07-21_11-20-13.sav.gz](https://github.com/user-attachments/files/16325115/Round-4-autosave_2024-07-21_11-20-13.sav.gz)

**Comment by repligator** [comment]
Further testing has produced interesting results.

A map with no enemies can be bombarded all day no issues, even if you accidentally damage your own units.

A map with a Princess (even if she has no units, and/or is friendly to the player) will experience the issue.

**Comment by repligator** [comment]
The total amount of artillery does _not_ have any affect. Slowdown will become noticeable after the 3rd round, and lock around the 4th.

If the artillery barrages are delayed for any reason (for example of they deploy at the start of the 5th round) the slowdown and lock will occur 3/4 turns after the barrage begins.

**Comment by repligator** [comment]
Using the memory display in IDEA, java.util.LinkedList$Node is increasing rapidly. All the instances I can see are `Artillery missed here on round 1, fired by Human` . 

![Screenshot_20240721_163403](https://github.com/user-attachments/assets/baf93666-1f33-489b-b359-25a45cb21109)
![Screenshot_20240721_163650](https://github.com/user-attachments/assets/12ac904c-534c-4f63-8040-fa1460b5a434)

**Comment by repligator** [comment]
Tested with bombers. It seems I can bomb all day without issue. The little explosions on the hexes disappear though...

**Comment by Sleet01** [comment]
Okay, this is likely related to the initial pass at adding artillery and bomb drift markers, but the good news is it should already be fixed in the update PR: #5762 
This would also explain why I haven't seen any issues: I tweaked this a couple days ago to fix what looked like a graphical glitch.

The root cause is a check to prevent these SpecialHexDisplays from being deleted when updates from other clients would replace the entire SHD hashtable.
But currently there's no check that a given ARTILLERY_MISS SHD _isn't_ already in the new HashTable, leading to unlimited growth in every client's copy of the table.

The more players or bots in a game, the faster this would lead to an out-of-memory issue.  But I'm surprised that the impact after only a few turns is so noticeable: each instance should be very lightweight, unless it somehow stores a copy of the graphic internally...

**Comment by Sleet01** [comment]
> Tested with bombers. It seems I can bomb all day without issue. The little explosions on the hexes disappear though...

There's a Client option that determines whether specific markers appear, in the Board tab - they may be turned off there.
But I also added an explicit cleanup pass for the Bomb-related markers, as otherwise they would stay forever.
I set the cleanup pass to occur immediately after the new phase display, so you have a phase to see the markers without attack arrows and such all over them.

There's an argument to be made for only cleaning them up after the movement phase, though: since Artillery attack markers persist until the next indirect phase, we could leave the bomb markers until the next Attack phase.
I just thought it would be too busy.
